StagedRegionValueAggregator and RVAState interfaces

catoverdrive opened the pull request:
(pulling out bits from #6480)

This defines the StagedRegionValueAggregator and RVAState interfaces, implementing ArrayElementsAggregator and PrevNonNullAggregator as examples.

Builds on #6500 and #6499. (for clarity, I've put all the changes from those PRs into the first commit.)

patrick-schultz commented on hail/src/main/scala/is/hail/expr/ir/agg/ArrayElementRegionValueAggregator.scala:
```diff
@@ -0,0 +1,269 @@
+          s.copyFrom(initStateAddress(i)) },
+        container.addState(srvb2),
+        srvb2.advance()),
+      typ.setFieldPresent(region, off, 1),
```
I'm missing where the init states get initialized. Shouldn't there be a `typ.setFieldPresent(region, off, 0)` somewhere?

catoverdrive replied:
There's basically two steps to the state initialization---
1. The init states get initialized with their respective init ops. The array of agg states is set to missing, since we don't necessarily know the length of the array.
2. We acquire the length of the array, and copy the init states into an array of that length. That's why the `setFieldPresent` is there---we expect the entire tuple to be present and initialized, but the first field to be missing, and we need to toggle that.

This is set up because the ArrayAggElements node provides us an optional `knownLength` argument---if it exists, then we'll initialize the length in the init op, but if it doesn't, we need to pick up the length from the first array that we see during the seqOp.

patrick-schultz replied:
Okay, I think I see. So the init states are never referenced after the second initialization step? Where does the first step happen?

catoverdrive replied:
The first step happens in the initOp.

catoverdrive commented on hail/src/main/scala/is/hail/expr/ir/agg/RVAState.scala:
```diff
@@ -0,0 +1,103 @@
+  def unserialize(codec: CodecSpec): Code[InputBuffer] => Code[Unit]
+
+  def er: EmitRegion = EmitRegion(mb, region)
+  def using(definition: Code[Region])(f: => Code[Unit]): Code[Unit] =
```
I don't think it does---this is one of those things that I thought might have to happen initially and then realized I probably didn't need but forgot to take out.

patrick-schultz replied:
Not a big deal, I wouldn't spend time changing it now. Just making sure I'm not missing something.

catoverdrive commented on hail/src/main/scala/is/hail/expr/ir/agg/StagedRegionValueAggregator.scala:
```diff
@@ -0,0 +1,39 @@
+
+  def createState(mb: EmitMethodBuilder): State
+
+  def initOp(state: State, init: Array[RVAVariable], dummy: Boolean): Code[Unit]
```
I was having a hard time with the type signature of this function---wanted to be able to call initOp on a generic RVAState and cast to the correct state for the aggregator, but didn't necessarily want to have to put like 
```
s2 = state.asInstanceOf[State]
```
in every method. So I defined two versions of every method, e.g.
```
def initOp(state: State, init: Array[RVAVariable], dummy: Boolean): Code[Unit]
def initOp(state: RVAState, init: Array[RVAVariable]): Code[Unit] = 
    initOp(state.asInstanceOf[State], init, true)
```
in which the second one is the one that's called externally. The dummy variable was necessary since `State <: RVAState`. Wasn't sure of the right way to do this.

patrick-schultz commented on hail/src/test/scala/is/hail/expr/ir/AggregatorsSuite2.scala:
```diff
@@ -0,0 +1,243 @@
+      ArrayRange(0, 15, 1), elt.name,
+      If(invoke("%", elt, 2).ceq(0), NA(row.typ), row))
+
+//    val arrayToAgg = MakeArray(FastIndexedSeq(), TArray(row.typ))
```
delete?

patrick-schultz commented on hail/src/test/scala/is/hail/expr/ir/AggregatorsSuite2.scala:
```diff
@@ -0,0 +1,243 @@
+
+    Region.scoped { region =>
+      val offset = f(0)(region)
+      println(SafeRow(PTuple(FastIndexedSeq(row.typ.physicalType)), region, aggf(0)(region, offset)))
```
Replace println with an assertion? (In all three tests)

patrick-schultz commented on hail/src/main/scala/is/hail/expr/ir/agg/ArrayElementRegionValueAggregator.scala:
```diff
@@ -0,0 +1,269 @@
+    })
+}
+
+class ArrayElementRegionValueAggregator2(nestedAggs: Array[StagedRegionValueAggregator]) extends StagedRegionValueAggregator {
```
What does this do? Is it supposed to have unimplemented methods?

catoverdrive replied:
(for the sake of logging this conversation; this preserves the same structure as the old ArrayElementsAggregator, and I'll rename these accordingly.

The first one contains an initOp that initializes the state, and a seqOp that takes the array length and checks it.

The second one doesn't have an initOp; its seqOp takes two arguments: the index of the agg state array to update, and a second argument of type TVoid that defines the seqOp to perform on it. Since the way it's extracted enforces that the seq op is always structured like:
```
Begin(
  SeqOp(i, Seq(len), agg1),
  ArrayFor(array, elt,
    SeqOp(i, Seq(idx, nestedSeqOp), agg2)))
```

it should never need to construct a state. I'll rename the classes and errors to be more descriptive.

catoverdrive commented:
@patrick-schultz I addressed your comments in the second commit and added sum/count aggregators. I also refactored the tests to test them; they're otherwise still the same.
